Data/ByteString/Internal.hs

```diff
@@ -785,8 +829,69 @@ isSpaceChar8 :: Char -> Bool
 isSpaceChar8 = isSpaceWord8 . c2w
 {-# INLINE isSpaceChar8 #-}
 
+------------------------------------------------------------------------
+
+-- | The type of exception raised by 'overflowError'
+-- and on failure by overflow-checked arithmetic operations.
+newtype SizeOverflowException
+  = SizeOverflowException String
+
+instance Show SizeOverflowException where
+  show (SizeOverflowException err) = err
+
+instance Exception SizeOverflowException
+
+-- | Raises a 'SizeOverflowException',
+-- with a message using the given function name.
 overflowError :: String -> a
-overflowError fun = error $ "Data.ByteString." ++ fun ++ ": size overflow"
+overflowError fun = throw $ SizeOverflowException msg
+  where msg = "Data.ByteString." ++ fun ++ ": size overflow"
+
+-- | Add two non-negative numbers.
+-- Calls 'overflowError' on overflow.
+checkedAdd :: String -> Int -> Int -> Int
+checkedAdd fun x y
+  | r >= 0    = r
+  | otherwise = overflowError fun
+  where r = assert (min x y >= 0) $ x + y
+{-# INLINE checkedAdd #-}
+
+-- | Multiplies two non-negative numbers.
+-- Calls 'overflowError' on overflow.
+checkedMultiply :: String -> Int -> Int -> Int
+checkedMultiply fun !x@(I# x#) !y@(I# y#) = assert (min x y >= 0) $
+#if TIMES_INT_2_AVAILABLE
+  case timesInt2# x# y# of
+    (# 0#, _, result #) -> I# result
+    _ -> overflowError fun
+#else
+  case timesWord2# (int2Word# x#) (int2Word# y#) of
+    (# hi, lo #) -> case or# hi (uncheckedShiftRL# lo shiftAmt) of
+      0## -> I# (word2Int# lo)
+      _   -> overflowError fun
+  where !(I# shiftAmt) = finiteBitSize (0 :: Word) - 1
+#endif
+
+
+_toIntegralSized :: (Integral a, Integral b, Bits a, Bits b) => a -> Maybe b
+{-# INLINE _toIntegralSized #-}
+-- stupid hack to make sure specialized versions of 'toIntegralSized'
+-- are generated when the checkedToInt RULES fire
+_toIntegralSized = inline toIntegralSized
+
+checkedToInt :: Integral t => t -> Maybe Int
+{-# RULES
+   "checkedToInt/Int"  checkedToInt = _toIntegralSized :: Int  -> Maybe Int
+ ; "checkedToInt/Word" checkedToInt = _toIntegralSized :: Word -> Maybe Int
+#-}
+{-# NOINLINE [1] checkedToInt #-}
```

Comment:
I appreciate the effort, but I have mixed feelings about the complexity of the whole thing. I'm not sure about RULES in general (@sjakobi could you please take a lead on this?), but I would certainly suggest to avoid playing with stage restrictions `[1]` - it's such a can of worms.